### ISSUE TITLE
Linking against `-lz` seems missing

### ISSUE THREAD

**Issue by merkys** [samuelflores/MMB#26]
Recently [an error has been reported](https://bugs.debian.org/1086362) on Debian builds:
```
[100%] Linking CXX executable MMB
/usr/bin/cmake -E cmake_link_script CMakeFiles/MMB_EXEC.dir/link.txt --verbose=1
/usr/bin/c++ -D BuildNtC -D USE_OPENMM -g -O2 -ffile-prefix-map=/<<PKGBUILDDIR>>=. -fstack-protector-strong -fstack-clash-protection -Wformat -Werror=format-security -fcf-protection -O0 -fvisibility=hidden -Wall -pedantic -O3 -DNDEBUG -Wl,-z,relro CMakeFiles/MMB_EXEC.dir/src/RNABuilder.cpp.o -o MMB   -L/<<PKGBUILDDIR>>/SYSTEM  -Wl,-rpath,/<<PKGBUILDDIR>>/SYSTEM:/<<PKGBUILDDIR>>/obj-x86_64-linux-gnu: libMMB.so.4.0 /usr/lib/x86_64-linux-gnu/libSimTKmolmodel.so.3.1 /usr/lib/x86_64-linux-gnu/libSimTKsimbody.so.3.7 /usr/lib/x86_64-linux-gnu/libSimTKmath.so.3.7 /usr/lib/x86_64-linux-gnu/libSimTKcommon.so.3.7 /usr/lib/x86_64-linux-gnu/libblas.so /usr/lib/x86_64-linux-gnu/liblapack.so /usr/lib/x86_64-linux-gnu/libblas.so /usr/lib/x86_64-linux-gnu/liblapack.so -lpthread -lrt -ldl -lm
/usr/bin/ld: libMMB.so.4.0: undefined reference to `gzerror'
/usr/bin/ld: libMMB.so.4.0: undefined reference to `gzgetc'
/usr/bin/ld: libMMB.so.4.0: undefined reference to `gzgets'
/usr/bin/ld: libMMB.so.4.0: undefined reference to `gzopen'
/usr/bin/ld: libMMB.so.4.0: undefined reference to `gzread'
/usr/bin/ld: libMMB.so.4.0: undefined reference to `gzungetc'
/usr/bin/ld: libMMB.so.4.0: undefined reference to `gzclose_r'
/usr/bin/ld: libMMB.so.4.0: undefined reference to `gzbuffer'
/usr/bin/ld: libMMB.so.4.0: undefined reference to `gzeof'
collect2: error: ld returned 1 exit status
```
To me it seems that adding `-lz` to `LDFLAGS` should fix the issue, although I have not tried yet.

**Comment by samuelflores** [comment]
Hi Andrius,

I usually compile using cmake. wouldn't it be easier to follow that ? I
don't think I've gotten this error that way.

I wonder if these libraries are required by MMB. I wonder if maybe it is
only that gemmi is not being installed properly? gemmi has gz.hpp which is
the only place that any gz operations are being done.

Sam

**Comment by MadCatX** [comment]
MMB is already picking up zlib during configuration phase and tries to link against it. See:
https://github.com/samuelflores/MMB/blob/4e5984cfcae40306897234774433fbf58e131044/CMakeLists.txt#L72
https://github.com/samuelflores/MMB/blob/4e5984cfcae40306897234774433fbf58e131044/CMakeLists.txt#L280
https://github.com/samuelflores/MMB/blob/4e5984cfcae40306897234774433fbf58e131044/CMakeLists.txt#L335

This would have become an issue long time ago if we really weren't linking against zlib. When linking fails like this, could you check that zlib was detected correctly by CMake and that the ${ZLIB_LIBRARIES} variable is sensible?

**Comment by MadCatX** [comment]
A proper indication that zlib was correctly are `ZLIB_LIBRARIES` and `ZLIB_INCLUDE_DIRS` (mind the plural form of the variable names) set to sensible values and `ZLIB_FOUND` set to true. `ZLIB_LIBRARY_RELEASE` is a variable that is set internally by FindZLIB and probably should not be used outside of that CMake module. Is there anything in your build process that might be setting the `ZLIB_LIBRARY` variable (singular) before CMake calls `find_package(zlib)`?

**Comment by merkys** [comment]
Right, it seems that in my case CMake does not populate `ZLIB_LIBRARIES` and `ZLIB_INCLUDE_DIRS`.

**Comment by merkys** [comment]
OK, this is not a problem with Debian, the variables I cited are indeed set, I just did not look into the right place. What seems to be the problem, is missing `PRIVATE ${ZLIB_LIBRARIES}` in the following segment:
https://github.com/samuelflores/MMB/blob/4e5984cfcae40306897234774433fbf58e131044/CMakeLists.txt#L296-L299

Without it, the executable `${MMB_SHARED_TARGET}` is not linked against ZLIB, thus ZLIB's symbols are left undefined. After adding `PRIVATE ${ZLIB_LIBRARIES}` locally I no longer get the build failure.

**Comment by samuelflores** [comment]
I am so pleased with your support.

tack

Sam

**Comment by MadCatX** [comment]
> OK, this is not a problem with Debian, the variables I cited are indeed set, I just did not look into the right place. What seems to be the problem, is missing `PRIVATE ${ZLIB_LIBRARIES}` in the following segment:
> 
> https://github.com/samuelflores/MMB/blob/4e5984cfcae40306897234774433fbf58e131044/CMakeLists.txt#L296-L299
> 
> Without it, the executable `${MMB_SHARED_TARGET}` is not linked against ZLIB, thus ZLIB's symbols are left undefined. After adding `PRIVATE ${ZLIB_LIBRARIES}` locally I no longer get the build failure.

This is confusing. The error that you reference in the issue points to the MMB executable failing to link. The python wrapper you mention is a different thing, although both MMB and the wrapper should fail to link in the same way if the issue was with libMMB. Are you sure you are getting the same build error as the one that you originally referenced? It also seems to me that libMMB.so itself builds successfully. If that is, indeed, the case, could you run `ldd` on the binary so that we can see what libraries it is actually linking against?

**Comment by samuelflores** [comment]
The python wrapper is not being used. We made it to make a Chimera GUI
plugin, but that is not working.

**Comment by merkys** [comment]
You are right, I have cited the wrong segment. I have opened PR #27 which fixes the build issue.

**Comment by samuelflores** [comment]
Now the problem seems to be occurring down at the molmodel level. I get the following on the "samuelflores/mmb-ubuntu" docker image, after updating to. latest gemmi and molmodel. Maybe you are building molmodel to a specific release?

root@0f646d5f78ff:/github/molmodel/build# make    
[ 25%] Built target SimTKmolmodel
[ 26%] Linking CXX executable TestAtomTargetLocations
/usr/bin/ld: ../libSimTKmolmodel.so.3.1: undefined reference to `gzerror'
/usr/bin/ld: ../libSimTKmolmodel.so.3.1: undefined reference to `gzgetc'
/usr/bin/ld: ../libSimTKmolmodel.so.3.1: undefined reference to `gzgets'
/usr/bin/ld: ../libSimTKmolmodel.so.3.1: undefined reference to `gzopen'
/usr/bin/ld: ../libSimTKmolmodel.so.3.1: undefined reference to `gzread'
/usr/bin/ld: ../libSimTKmolmodel.so.3.1: undefined reference to `gzungetc'
/usr/bin/ld: ../libSimTKmolmodel.so.3.1: undefined reference to `gzclose_r'
/usr/bin/ld: ../libSimTKmolmodel.so.3.1: undefined reference to `gzbuffer'
/usr/bin/ld: ../libSimTKmolmodel.so.3.1: undefined reference to `gzeof'
collect2: error: ld returned 1 exit status
make[2]: *** [tests/CMakeFiles/TestAtomTargetLocations.dir/build.make:105: tests/TestAtomTargetLocations] Error 1
make[1]: *** [CMakeFiles/Makefile2:1030: tests/CMakeFiles/TestAtomTargetLocations.dir/all] Error 2
make: *** [Makefile:146: all] Error 2
root@0f646d5f78ff:/github/molmodel/build#

**Comment by samuelflores** [comment]
never mind, I guess i just needed to build cleaner from the bottom up.

**Comment by merkys** [comment]
In case you are interested in exact versions of build dependencies, they can be found in [Debian build log](https://people.debian.org/~sanvila/build-logs/202410/macromoleculebuilder_4.0.0+dfsg-4_amd64-20241030T084237.918Z). Probably the most relevant part:
```
Get:158 http://deb.debian.org/debian sid/main amd64 tao-pegtl-dev all 3.2.7-1 [69.4 kB]
Get:159 http://deb.debian.org/debian sid/main amd64 gemmi-dev amd64 0.6.5+ds-2 [1139 kB]
Get:160 http://deb.debian.org/debian sid/main amd64 libblas3 amd64 3.12.0-3+b1 [152 kB]
Get:161 http://deb.debian.org/debian sid/main amd64 libblas-dev amd64 3.12.0-3+b1 [161 kB]
Get:162 http://deb.debian.org/debian sid/main amd64 libgfortran5 amd64 14.2.0-7 [837 kB]
Get:163 http://deb.debian.org/debian sid/main amd64 liblapack3 amd64 3.12.0-3+b1 [2450 kB]
Get:164 http://deb.debian.org/debian sid/main amd64 liblapack-dev amd64 3.12.0-3+b1 [4990 kB]
Get:165 http://deb.debian.org/debian sid/main amd64 libopenmm8.1 amd64 8.1.2+dfsg-1 [1058 kB]
Get:166 http://deb.debian.org/debian sid/main amd64 libopenmm-dev amd64 8.1.2+dfsg-1 [345 kB]
Get:167 http://deb.debian.org/debian sid/main amd64 libseqan2-dev all 2.4.0+dfsg-16 [1216 kB]
Get:168 http://deb.debian.org/debian sid/main amd64 libxfixes-dev amd64 1:6.0.0-2+b2 [22.3 kB]
Get:169 http://deb.debian.org/debian sid/main amd64 libxi-dev amd64 2:1.8.2-1 [241 kB]
Get:170 http://deb.debian.org/debian sid/main amd64 libxmu6 amd64 2:1.1.3-3+b3 [58.8 kB]
Get:171 http://deb.debian.org/debian sid/main amd64 libxmu-headers all 2:1.1.3-3 [68.5 kB]
Get:172 http://deb.debian.org/debian sid/main amd64 libxmu-dev amd64 2:1.1.3-3+b3 [65.2 kB]
Get:173 http://deb.debian.org/debian sid/main amd64 libsimbody3.7 amd64 3.7+dfsg-3+b1 [3363 kB]
Get:174 http://deb.debian.org/debian sid/main amd64 libsimbody-dev amd64 3.7+dfsg-3+b1 [2728 kB]
Get:175 http://deb.debian.org/debian sid/main amd64 libsimtkmolmodel3.1t64 amd64 3.1.0-4.1+b3 [667 kB]
Get:176 http://deb.debian.org/debian sid/main amd64 libsimtkmolmodel-dev amd64 3.1.0-4.1+b3 [1864 kB]
Get:177 http://deb.debian.org/debian sid/main amd64 libsimtkmolmodel-plugins amd64 3.1.0-4.1+b3 [19.6 kB]
Get:178 http://deb.debian.org/debian sid/main amd64 libstb0t64 amd64 0.0~git20240715.f7f20f39fe4f+ds-1 [220 kB]
Get:179 http://deb.debian.org/debian sid/main amd64 libstb-dev amd64 0.0~git20240715.f7f20f39fe4f+ds-1 [654 kB]
Get:180 http://deb.debian.org/debian sid/main amd64 zlib1g-dev amd64 1:1.3.dfsg+really1.3.1-1+b1 [920 kB]
```
